### PR TITLE
[midi] Fix center-64 MIDI VFO tuning jumps (CTR2-Midi)

### LINKED ISSUE #4096: VFO tuning (just one bump) results in a jump greater than the step of 100 using CTR2-MIDI controller
[support-bundle-20260707-071535.zip](https://github.com/user-attachments/files/29747207/support-bundle-20260707-071535.zip)

I'm experiencing an issue with AetherSDR, on macOS
for FlexRadio transceivers Flex 6400 (https://github.com/aethersdr/AetherSDR).

My system:
- AetherSDR version: 26.6.4
- Qt version: 6.11.1
- OS: macOS

Here is my issue:

1. Using CTR2-MIDI as my VFO/tuning interface ( https://ctr2.lynovation.com/ctr2-midi/ ).
2. Using this has worked well in a prior version of AetherSDR several months ago . . .  - I recently updated to the current version 26.6.4 and now tuning the VFO with the knob (just one bump) results in a jump greater than the step of 100 I have it set to (see attached image). 
3. There is no issue when using SmartSDR . . . only using AetherSDR

Image

### PR DESCRIPTION
**Pull request by jensenpat**
## Summary

- Detect center-64 relative MIDI CC encoding for learned VFO Tune Knob bindings, so CTR2-MIDI values `65` and `63` produce one clockwise/counter-clockwise step.
- Preserve the established two's-complement `1`/`127` behavior for existing relative encoders by locking the detected encoding per binding.
- Add a focused decoder regression test and an RX-only `midi cc` automation verb for repeatable controller-path proof.

## Root cause

MIDI Learn marks every VFO Control Change binding as relative. The explicit-relative path then unconditionally decoded values as two's complement, so CTR2-MIDI's center-64 detents (`65`/`63`) became `-63`/`+63`. The center-64 fallback added in #2994 was guarded by `!binding.relative`, making it unreachable for learned VFO bindings.

The reporter's follow-up log shows the exact signature: 14.300 MHz moved to 14.363 MHz and then 14.237 MHz in 63-step increments.

## Proof

Built the macOS app and drove the same learned VFO relative decoder through the agent automation bridge on a connected FLEX-8400M (firmware 4.2.18.41174), with the configured 100 Hz tuning step:

| Action | Active-slice frequency |
|---|---:|
| Before | 14.100000 MHz |
| `midi cc 65` | 14.100100 MHz |
| `midi cc 63` | 14.100000 MHz |

Both detents moved exactly one 100 Hz step in opposite directions, the starting frequency was restored, and the final radio snapshot reported `transmitting:false`.

VFO at 14.100100 MHz after one CTR2 clockwise detent

## Tests

- `cmake --build build --target AetherSDR -j8`
- `midi_relative_cc_decoder_test`
- `midi_settings_test`
- `bridge_docs_check`
- `tools/check_engine_boundary.py --strict` (zero blocking findings)
- `tools/check_a11y.py` (existing warning-only findings)

Fixes #4096

👨🏼‍💻 Generated with OpenAI Codex (GPT-5.6 Sol) and tested by @jensenpat

### CODE REVIEW

File: src/core/MidiRelativeCcDecoder.h [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+// two's-complement 1/127 convention unless the first directional value is the
+// distinctive 63/65 pair used by center-64 controllers such as CTR2-MIDI.
+inline MidiRelativeCcDecodeResult decodeMidiRelativeCc(
+    int value, MidiRelativeCcEncoding currentEncoding)
+{
+    MidiRelativeCcEncoding encoding = currentEncoding;
+    if (encoding == MidiRelativeCcEncoding::Undetermined) {
+        if (value == 63 || value == 65) {
+            encoding = MidiRelativeCcEncoding::Center64;
+        } else if (value != 64) {
```
[the remark this answers is not in the document]
Comment (skerker) [reply src/core/MidiRelativeCcDecoder.h]: @jensenpat  Saw the help wanted - not sure if you are looking for hardware testing?  I was already in the process of testing my CTR2-MIDI this week, when I came across this PR.    I've got a CTR2-MIDI on the bench and ran it through your midi cc bridge path: the fix works — slow first click locks center-64, and both slow and fast detents tune the right direction and amount (65→+100 Hz, 72→+800 Hz, and back).

I confirmed the auto-detect edge case the bot flagged is real (a fast first value like 72 does lock two's-complement), but on this controller it's hard to hit in practice — the firmware's speed-tuning always sends ±1 first from rest, so a real first touch lands on 63/65 every time.

One extra thing to note: if I change the knob's MIDI Type (WheelA→WheelB) without re-Learning, it tunes in an unexpected way — the binding stays locked to the encoding detected at the previous Learn, so a re-Learn is needed after any MIDI Type change for it to decode correctly.   As I understand it, this PR decodes the two relative wheel encodings — WheelA (center-64) and WheelB (two's-complement) — so the user needs the knob set to WheelA or WheelB for VFO tuning. The CTR2 offers other MIDI Types too (sliders, a button, WheelB-r) that I haven't tested.

Happy to share the capture logs and/or test further, if that's the help you are looking for?

— authored by agent (Claude Code) on behalf of @skerker

**Review by NF0T — DISMISSED** [review]
Correct root-cause diagnosis and the right general shape — auto-detecting center-64 vs. two's-complement from the first directional value, with a per-binding lock, closes the reported CTR2-MIDI bug cleanly, and the isolated decoder header + focused test is nice work. CI is green, commit is signed, and the live bridge proof plus @skerker's independent hardware test on real CTR2-MIDI both confirm the fix works for the reported case.

Two issues found beyond what's already been discussed here:

**1. The automation injection path and a real controller can corrupt each other's encoding lock.** `MidiBinding::key()` is `(channel<<16)|(msgType<<8)|number` — it doesn't include `paramId`. `injectVfoCcForAutomation()` builds a synthetic binding with `channel=0, CC, number=0` (key=0) to read/write `m_relativeCcEncodings`. If a real user's VFO-tune-knob binding is also learned on MIDI channel 1 (internal 0) / CC 0 — a fully reachable combination — its key is the same. Running the `midi cc` automation verb with a value outside {63,64,65} in the same process lifetime as that real controller can permanently mis-lock its encoding, producing a large wrong-direction jump on every subsequent real detent. Suggest either a reserved sentinel key the real `key()` function can never produce, or looking up the actual `rx.tuneKnob` binding's real key instead of hardcoding channel/CC 0/0.

**2. Editing any unrelated MIDI binding silently wipes the VFO knob's already-detected encoding.** `rebuildIndex()`/`clearBindings()` now unconditionally clear all of `m_relativeCcEncodings`. `MidiMappingDialog.cpp` (untouched by this PR) calls `rebuildIndex()` from three existing paths that have nothing to do with the VFO binding: toggling Invert on any binding, toggling Relative on any binding, and deleting any single binding. So a user with a correctly-locked, working CTR2-MIDI VFO knob who opens the Mapping dialog to adjust something unrelated (e.g. an AF-Gain slider's Invert flag) gets their VFO encoding silently reset to `Undetermined`, with no visible indication — reopening the already-discussed "wrong first sample" window far more often than once at device connect. This changes the risk profile of the fast-first-detent edge case both bot reviews characterized as low-practical-risk, since it can now recur on any unrelated Mapping-dialog interaction, not just once per session. Suggest scoping the clear to just the affected binding's key rather than the whole map (or skip the clear entirely for edits that don't touch the VFO binding's own key).

Non-blocking: the header comment describes only the happy path — worth a line noting the "locks from first sample, no re-detection" limitation now that finding #2 shows it's reachable more than once per session.

Happy to re-review quickly once these are addressed — the core approach is right, these are both reachability/robustness gaps rather than problems with the detection heuristic itself.

File: src/core/MidiRelativeCcDecoder.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (jensenpat) [reply src/core/MidiRelativeCcDecoder.h]: @skerker Thank you for testing and confirmation. Would you mind opening a separate issue on the re-learn behavior (and/or PR) since you have the wheel and can test please?  Please feel free to pick that PR up if you have cycles.

[the remark this answers is not in the document]
Comment (skerker) [reply src/core/MidiRelativeCcDecoder.h]: @jensenpat   Happy to help  - I'll submit a separate issue on the WheelA / WheelB re-learn behavior and work on the PR.

File: src/core/MidiRelativeCcDecoder.h [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+            encoding = MidiRelativeCcEncoding::TwosComplement;
+        }
+    }
```
Comment (ten9876) [inline src/core/MidiRelativeCcDecoder.h:37]: 🔴 **Blocker — locks the encoding on an ambiguous first value.** The two encodings' *unit* detents are disjoint (1/127 vs 63/65), but `else if (value != 64)` locks two's-complement on **any** other first sample. A center-64 controller whose first move isn't a single click (e.g. a quick spin → value 70) locks two's-complement, and every later 63/65 detent then decodes to ∓63 — the #4096 jumps, made permanent with no recovery short of re-learning. Proven by compiling this header standalone.

Fix: only lock on a *definitive* discriminator; defer (delta 0, stay `Undetermined`) on an ambiguous first sample until a real unit detent arrives. Verified this still passes all existing tests.

```suggestion
    if (encoding == MidiRelativeCcEncoding::Undetermined) {
        // Only lock on a *definitive* single-detent discriminator. The two
        // encodings' unit values are disjoint (1/127 vs 63/65); any other
        // first sample (e.g. a fast first turn) is ambiguous and must NOT
        // commit an encoding — otherwise a center-64 controller whose first
        // move isn't a unit detent locks two's-complement and every later
        // 63/65 detent decodes to ∓63 (the #4096 jumps, made permanent).
        if (value == 63 || value == 65) {
            encoding = MidiRelativeCcEncoding::Center64;
        } else if (value == 1 || value == 127) {
            encoding = MidiRelativeCcEncoding::TwosComplement;
        } else {
            return {0, encoding};  // stay Undetermined until a unit detent
        }
    }
```

File: src/core/MidiControlManager.cpp [hunk 3]
```diff
@@ -371,21 +389,13 @@ void MidiControlManager::onMidiMessage(int status, int data1, int data2,
...
-    // for the VFO would silently flip CCW pulses (data2=127, two's-complement
-    // -1) to CW for users with existing two's-complement encoders bound to
-    // the VFO knob.  Binary-mode encoders (data2 ∈ {0, 127}) instead fall
-    // through to the backward-compat Tier 1 below, which handles them
-    // without requiring the relative flag.
+    // Learned VFO bindings auto-detect the two common encodings from the first
+    // directional value: 1/127 remains two's-complement, while the distinctive
+    // 63/65 pair selects center-64. Other relative parameters retain the
+    // established two's-complement behavior. Binary-mode encoders fall through
+    // to the backward-compat Tier 1 below when not explicitly marked relative.
```
Comment (ten9876) [inline src/core/MidiControlManager.cpp]: 🟠 **This comment is inaccurate for learned bindings, and it hides an unsupported case.** `startLearn` always sets `binding.relative = isVfoTuneKnobParamId(...) && CC`, so every *newly learned* VFO CC binding takes the `binding.relative` branch — Tier 1 (guarded by `!binding.relative`) is never reached for it. A binary/Thetis (0/127) encoder learned on the VFO therefore does **not** fall through; it's decoded as two's-complement here (`127 → −1` reversed, `0 → 0` dead). Either the comment should say so, or binary encoders need real handling (they can't just reuse Tier 1, since `127` collides between binary-CW and two's-complement-−1).

```suggestion
    // Learned VFO bindings auto-detect the two common encodings from the first
    // *unit* detent: 1/127 → two's-complement, 63/65 → center-64. Other
    // relative parameters keep two's-complement. NOTE: learned VFO CC bindings
    // are always marked relative (startLearn), so Tier 1/2 below are NOT reached
    // for them — a binary/Thetis (0/127) encoder learned on the VFO is decoded
    // as two's-complement here (127 → −1). Genuine binary support needs its own
    // encoding, not the Tier-1 fall-through (127 collides with two's-comp −1).
```

File: src/gui/MainWindow_Controllers.cpp [hunk 4]
```diff
@@ -1292,6 +1292,27 @@ void MainWindow::refreshStreamDeckLabels()
...
+#ifdef HAVE_MIDI
+    if (!m_midiControl) {
+        return 1;
+    }
+
+    bool accepted = false;
+    const bool invoked = QMetaObject::invokeMethod(
+        m_midiControl, "injectVfoCcForAutomation", Qt::BlockingQueuedConnection,
+        Q_RETURN_ARG(bool, accepted), Q_ARG(int, value));
+    return invoked && accepted ? 0 : 1;
```
Comment (ten9876) [inline src/gui/MainWindow_Controllers.cpp:1320]: 🟠 **Unconditional `BlockingQueuedConnection` skips the codebase's same-thread guard.** The established idiom (e.g. `AutomationServer.cpp` ~1590) checks `!t || t == QThread::currentThread()` and only block-queues when genuinely cross-thread. It works today because MIDI lives on `m_extCtrlThread`, but it's a latent GUI hang: a same-thread caller would deadlock, and a dead/quitting ext-ctrl loop (teardown) would block the GUI forever. Match the idiom (needs `#include <QThread>` if not already present):

```suggestion
    // Only block-queue when m_midiControl is genuinely on another thread —
    // a same-thread BlockingQueuedConnection deadlocks, and a dead/quitting
    // ext-ctrl loop would hang the GUI (cf. AutomationServer.cpp ~1590).
    QThread* mgrThread = m_midiControl->thread();
    const Qt::ConnectionType conn =
        (!mgrThread || mgrThread == QThread::currentThread())
            ? Qt::DirectConnection : Qt::BlockingQueuedConnection;
    bool accepted = false;
    const bool invoked = QMetaObject::invokeMethod(
        m_midiControl, "injectVfoCcForAutomation", conn,
        Q_RETURN_ARG(bool, accepted), Q_ARG(int, value));
    return invoked && accepted ? 0 : 1;
```

File: tests/midi_relative_cc_decoder_test.cpp [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    ok &= expect(result.delta == -1,
+                 "two's-complement counter-clockwise pulse is preserved");
+
+    encoding = MidiRelativeCcEncoding::Undetermined;
+    result = decodeMidiRelativeCc(64, encoding);
+    ok &= expect(result.delta == 0, "center value is neutral before detection");
+    ok &= expect(result.encoding == MidiRelativeCcEncoding::Undetermined,
+                 "neutral value does not lock an encoding");
+
+    return ok ? 0 : 1;
```
Comment (ten9876) [inline tests/midi_relative_cc_decoder_test.cpp:78]: 🟡 **The test only exercises clean single detents — exactly the happy path that hides finding #1.** Add coverage for the ambiguous-first-value case so the regression can't slip back in:

```suggestion

    // Ambiguous first value must NOT lock an encoding (#4096 regression guard):
    // a center-64 controller whose first move isn't a unit detent would else
    // lock two's-complement and decode every later detent to ∓63.
    encoding = MidiRelativeCcEncoding::Undetermined;
    result = decodeMidiRelativeCc(70, encoding);
    encoding = result.encoding;
    ok &= expect(result.delta == 0, "ambiguous first value yields no step");
    ok &= expect(encoding == MidiRelativeCcEncoding::Undetermined,
                 "ambiguous first value does not lock an encoding");
    result = decodeMidiRelativeCc(65, encoding);
    encoding = result.encoding;
    ok &= expect(result.delta == 1 && encoding == MidiRelativeCcEncoding::Center64,
                 "first real detent still selects center-64 (+1)");

    return ok ? 0 : 1;
```

File: src/core/MidiControlManager.cpp [hunk 6]
```diff
@@ -468,6 +478,29 @@ void MidiControlManager::onMidiMessage(int status, int data1, int data2,
...
+        encoding = result.encoding;
+        delta = result.delta;
+    } else {
+        delta = relativeCcDelta(value);
+    }
+
+    if (binding.inverted) {
+        delta = -delta;
+    }
+    if (delta == 0) {
```
Comment (ten9876) [inline src/core/MidiControlManager.cpp:522]: 🟡 **Minor behavior change — `delta == 0` now skips the `paramValueChanged(paramId, 0.0f)` emit** that the old explicit-relative path emitted for `value == 0` on non-VFO relative params. Almost certainly harmless (and arguably cleaner), but flagging in case any consumer watches that signal for a zero pulse.

**Review by ten9876 — CHANGES_REQUESTED** [review]
## Review — center-64 MIDI VFO decode

Root-cause diagnosis is right and the center-64 math is correct for clean single detents. But the **encoding detector locks on the first non-64 value even when that value is ambiguous**, which lets the exact #4096 bug come back — permanently — for the very controller this PR targets. I compiled the decoder header in isolation to confirm:

```
CTR2 (center-64), first action = a quick spin → value 70 → locks TwosComplement
   then CW detent 65 → delta −63   (should be +1)
   then CCW detent 63 → delta +63  (should be −1)   ← #4096 63-step jumps, now permanent
```

Inline comments below carry ready-to-apply suggestions.

**Blocker:** #1 (ambiguous-first-value mis-lock). **Same-pass:** #2 (binary/Thetis 0/127 encoders + inaccurate comment), #3 (thread-guard idiom), #4 (test gap that hid #1). #5 is minor.

**Deliberately not re-raised** (already covered by @NF0T's `CHANGES_REQUESTED`, +1 to both):
- The `injectVfoCcForAutomation` `key()` collision with a real ch1/CC0 VFO binding — NF0T's finding #1.
- `rebuildIndex()`/`clearBindings()` wiping the detected VFO encoding on any unrelated Mapping-dialog edit — NF0T's finding #2. Worth noting this one **compounds my #1**: it reopens the ambiguous-first-sample window on every unrelated binding edit, not just once per session, which is exactly why I'd treat the mis-lock as a real blocker rather than a rare edge case.

All suggestions verified: the reworked decoder passes every existing test plus the new mis-detection guard.

**Review by ten9876 — DISMISSED** [review]
Approving with a follow-up commit (`a7d84c73`, signed) that resolves every review finding — mine and @NF0T's. Built clean; `midi_relative_cc_decoder_test` + `midi_settings_test` pass (incl. new regression cases).

**Fixed**
- **Decoder mis-lock (the #4096 regression).** `decodeMidiRelativeCc` locked two's-complement on *any* first non-64 value, so a center-64 controller whose first move wasn't a unit detent (e.g. a fast turn landing on 70) locked the wrong encoding and decoded every later 63/65 detent as ∓63 — permanently. Now only a definitive unit detent commits an encoding (`{1,127}`→two's-comp, `{63,65}`→center-64); any other first sample defers until one settles it. I'd confirmed the original bug by compiling the header; the new tests pin both directions.
- **`rebuildIndex()` wiped the whole encoding map (NF0T #2).** Toggling Invert/Relative on — or deleting — *any* unrelated binding reset a working VFO lock. Now prunes only removed keys; live bindings keep their encoding.
- **Automation `key()` collision (NF0T #1).** `injectVfoCcForAutomation` built a synthetic ch0/CC0 binding sharing `m_relativeCcEncodings[key 0]` with a real knob on MIDI ch1/CC0. Now decodes against a dedicated `m_automationCcEncoding` field, never per-binding state.
- **Threading.** The automation injection now applies the same-thread guard idiom instead of an unconditional `BlockingQueuedConnection`.
- **Comment corrected.** Learned VFO bindings are always relative, so Tier 1 is unreachable for them and a binary/Thetis (0/127) encoder is decoded here as two's-complement — genuine binary support filed as **#4402** (0/127 is ambiguous with two's-complement's own unit values, so it can't be auto-detected from the value alone).

**Deliberately kept:** the `delta==0` early-return in `dispatchRelativeCc`. On closer look it's *correct*, not a regression — value 0 is no detent, and the pre-PR path emitted a spurious 0.0 (decrease) pulse and inflated the acceleration event count for a non-movement.

Thanks @jensenpat — the center-64 diagnosis and the isolated decoder-header design were the right calls; this just closes the detection edge and the two reachability gaps.

**Review by ten9876 — APPROVED** [review]
Re-approving on the rebased head (`d37fa8ef`). Both of @NF0T's blocking findings are addressed in this revision:

- **#1 automation/real-controller collision** — `injectVfoCcForAutomation()` no longer builds a synthetic ch0/CC0 binding; it decodes against a dedicated `m_automationCcEncoding` field and never touches `m_relativeCcEncodings`, so there's no shared key to collide on.
- **#2 `rebuildIndex()` wiping the VFO lock** — it now prunes only keys whose binding was removed, so an unrelated Mapping-dialog edit (Invert/Relative toggle, or deleting another binding) preserves the detected encoding. `clearBindings()` stays clear-all, since it only runs when every binding is gone.

Plus the decoder mis-lock (the #4096 root cause), the thread-guard idiom, the corrected binary-encoder comment (follow-up #4402), and the ambiguous-first-value regression tests.

Rebased onto current main (verb registry auto-merged, `docs/automation-bridge.md` regenerated to 52 verbs); builds clean and `midi_relative_cc_decoder_test` / `midi_settings_test` / `bridge_docs_check` all pass. @NF0T — your two asks are done if you'd like to clear the review.
